File: TestLeftSliderDemoForOC/TestLeftSliderDemoForOC/LeftSliderManager/WHLeftSlideManager.m
//
//  WHLeftSlideManager.m
//  TestLeftSliderDemoForOC
//
//  Created by 邓伟浩 on 2020/6/1.
//  Copyright © 2020 邓伟浩. All rights reserved.
//

#import "WHLeftSlideManager.h"

/// 手势轻扫临界速度
CGFloat const LSLeftSlipCriticalVelocity = 800;
/// 左滑手势触发距离
CGFloat const LSLeftSlipLeftSlipPanTriggerWidth = 50;

@interface WHLeftSlideManager ()<UIViewControllerTransitioningDelegate, UIGestureRecognizerDelegate, UIViewControllerAnimatedTransitioning>

@property (nonatomic, weak) UIViewController *leftVC;
//@property (nonatomic, weak) UIViewController *mainVC;

/** 点击返回的遮罩view */
@property (nonatomic, strong) UIView *tapView;

/** present or dismiss */
@property (nonatomic, assign) BOOL isPresent;

/** 是否已经显示左滑视图 */
@property (nonatomic, assign) BOOL isShowLeft;

/** 是否在交互中 */
@property (nonatomic, assign) BOOL interactive;

/** 主VC是否跟随滑动 */
@property (nonatomic, assign) BOOL shouldMove;

/** 左滑视图宽度 */
@property (nonatomic, assign) CGFloat leftViewWidth;

/** 侧滑手势 */
@property (nonatomic, strong) UIPanGestureRecognizer *panGesture;


@end

@implementation WHLeftSlideManager

+ (instancetype)sharedManager {
    static WHLeftSlideManager *sharedWHLeftSlideManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedWHLeftSlideManager = [[self alloc] init];
    });
    return sharedWHLeftSlideManager;
}

#pragma mark - ———— 公共方法 ————

- (CGFloat)scaleFor:(CGFloat)width {
    return ((width)*(([[UIScreen mainScreen] bounds].size.width)/375.0f));
}

- (void)setLeftViewController:(UIViewController *)leftViewController mainViewController:(UIViewController *)mainViewController {
    [self setLeftViewController:leftViewController leftViewWidth:[self scaleFor:310] mainViewController:mainViewController shouldMove:YES];
}

- (void)setLeftViewController:(UIViewController *)leftViewController leftViewWidth:(CGFloat)leftViewWidth mainViewController:(UIViewController *)mainViewController shouldMove:(BOOL)shouldMove {
    self.leftVC = leftViewController;
    self.mainVC = mainViewController;
    self.leftViewWidth = leftViewWidth;
    self.shouldMove = shouldMove;
    
    if (self.mainVC.navigationController == nil) {
        [self.mainVC.view addSubview:self.tapView];
    } else {
        [self.mainVC.navigationController.view addSubview:self.tapView];
    }
    self.tapView.hidden = YES;
    
    self.leftVC.transitioningDelegate = self;
    // 侧滑手势
    [self.mainVC.view addGestureRecognizer:self.panGesture];
}

/** 显示菜单 */
- (void)showLeftMenuView {
    if (!self.isShowLeft) {
        self.leftVC.modalPresentationStyle = UIModalPresentationFullScreen;
        [self.mainVC presentViewController:self.leftVC animated:YES completion:nil];
    } else {
        [self dismissLeftMenuView];
    }
    
}

/** 取消显示菜单 */
- (void)dismissLeftMenuView {
    //    dispatch_async(dispatch_get_main_queue(), ^{
    [self.leftVC dismissViewControllerAnimated:YES completion:nil];
    //    });
}

#pragma mark - ———— 手势处理方法 ————
- (void)pan:(UIPanGestureRecognizer *)pan {
    // X轴偏移
    CGFloat offsetX = [pan translationInView:pan.view].x;
    // X轴速度
    CGFloat velocityX = [pan velocityInView:pan.view].x;
    
    CGFloat percent;
    if (self.isShowLeft) {
        // 坑点。千万不要超过1
        percent = MIN(-offsetX / self.leftViewWidth, 1);
    } else {
        percent = MIN(offsetX / self.leftViewWidth, 1);
    }
    
    switch (pan.state) {
        case UIGestureRecognizerStateBegan:
        {
            self.interactive = YES;
            if (self.isShowLeft) {
                [self dismissLeftMenuView];
            } else {
                [self showLeftMenuView];
            }
        }
            break;
        case UIGestureRecognizerStateChanged:
        {
            [self updateInteractiveTransition:percent];
        }
            break;
        case UIGestureRecognizerStateEnded:
        case UIGestureRecognizerStateCancelled:
        {
            self.interactive = NO;
            
            // 判断是否需要转场
            BOOL shouldTransition = NO;
            
            // 1.present时
            // 1.1 速度正方向，>800，则正向转场
            // 1.2 速度反向时，<-800，则反向转场
            // 1.3 速度正向<800 或者 速度反向>-800， 判断percent是否大于0.5
            if (!self.isShowLeft) {
                if (velocityX > 0) {
                    if (velocityX > LSLeftSlipCriticalVelocity) {
                        shouldTransition = YES;
                    } else {
                        shouldTransition = percent > 0.5;
                    }
                } else {
                    if (velocityX < -LSLeftSlipCriticalVelocity) {
                        shouldTransition = NO;
                    } else {
                        shouldTransition = percent > 0.5;
                    }
                }
            } else {
                if (velocityX < 0) {
                    if (velocityX < -LSLeftSlipCriticalVelocity) {
                        shouldTransition = YES;
                    } else {
                        shouldTransition = percent > 0.5;
                    }
                } else {
                    if (velocityX > LSLeftSlipCriticalVelocity) {
                        shouldTransition = NO;
                    } else {
                        shouldTransition = percent > 0.5;
                    }
                }
            }
            
            // 2.dismiss时
            // 2.1 速度正向，<-800，则正向转场
            // 2.2 速度反向，>800，则反向转场
            // 2.3 速度正向>-800 或者 速度反向<800，判断percent是否大于0.5
            if (shouldTransition) {
                [self finishInteractiveTransition];
            } else {
                [self cancelInteractiveTransition];
            }
        }
            break;
        default:
            break;
    }
}

#pragma mark - UIGestureRecognizerDelegate Methods
- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    if (self.isShowLeft) {
        return YES;
    }
    
    UIPanGestureRecognizer *panGesture = (UIPanGestureRecognizer *)gestureRecognizer;
    
    // 忽略起始点不在左侧触发范围内的手势
    CGFloat touchBeganX = [panGesture locationInView:panGesture.view].x;
    if (touchBeganX > LSLeftSlipLeftSlipPanTriggerWidth) {
        return NO;
    }
    
    // 忽略反向手势
    CGPoint translation = [panGesture translationInView:panGesture.view];
    if (translation.x <= 0) {
        return NO;
    }
    
    return YES;
}

#pragma mark - ———— UIViewControllerTransitioningDelegate代理方法 ————
- (nullable id <UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source {
    self.isPresent = YES;
    return self;
}

- (nullable id <UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed {
    self.isPresent = NO;
    return self;
}

- (id<UIViewControllerInteractiveTransitioning>)interactionControllerForDismissal:(id<UIViewControllerAnimatedTransitioning>)animator {
    return self.interactive ? self : nil;
}

- (id<UIViewControllerInteractiveTransitioning>)interactionControllerForPresentation:(id<UIViewControllerAnimatedTransitioning>)animator {
    return self.interactive ? self : nil;
}

#pragma mark - ———— UIViewControllerAnimatedTransitioning代理方法 ————
- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
    return .3f;
}

- (void)animateTransition:(id <UIViewControllerContextTransitioning>)transitionContext {
    if (self.isPresent) {
        UIViewController *toVC = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
        UIViewController *fromVC = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
        
        UIView *toView = nil;
        UIView *fromView = nil;
        
        if ([transitionContext respondsToSelector:@selector(viewForKey:)]) {
            fromView = [transitionContext viewForKey:UITransitionContextFromViewKey];
            toView = [transitionContext viewForKey:UITransitionContextToViewKey];
        } else {
            fromView = fromVC.view;
            toView = toVC.view;
        }
        
        UIView *containerView = [transitionContext containerView];
        [containerView addSubview:toView];
        
        toView.frame = CGRectMake(-self.leftViewWidth, 0, self.leftViewWidth, containerView.frame.size.height);
        
        [self.tapView.superview bringSubviewToFront:self.tapView];
        self.tapView.hidden = NO;
        
        // 动画block
        void(^animateBlock)(void) = ^{
            toView.frame = CGRectMake(0, 0, self.leftViewWidth, toView.frame.size.height);
            if (self.shouldMove) {
                fromView.frame = CGRectMake(self.leftViewWidth, 0, fromView.frame.size.width, fromView.frame.size.height);
            }
            self.tapView.alpha = 1.f;
        };
        
        // 动画完成block
        void(^completeBlock)(void) = ^{
            if ([transitionContext transitionWasCancelled]) {
                [transitionContext completeTransition:NO];
            } else {
                [transitionContext completeTransition:YES];
                [containerView addSubview:fromView];
                [containerView sendSubviewToBack:fromView];
                self.isShowLeft = YES;
            }
        };
        
        if (self.interactive) {
            // 呵呵🙃
            [UIView animateWithDuration:[self transitionDuration:transitionContext] delay:0 options:UIViewAnimationOptionCurveLinear animations:^{
                animateBlock();
            } completion:^(BOOL finished) {
                completeBlock();
            }];
        } else {
            [UIView animateWithDuration:[self transitionDuration:transitionContext] animations:^{
                animateBlock();
            } completion:^(BOOL finished) {
                completeBlock();
            }];
        }
    } else {
        UIViewController *toVC = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
        UIViewController *fromVC = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
        
        UIView *toView = nil;
        UIView *fromView = nil;
        
        if ([transitionContext respondsToSelector:@selector(viewForKey:)]) {
            fromView = [transitionContext viewForKey:UITransitionContextFromViewKey];
            toView = [transitionContext viewForKey:UITransitionContextToViewKey];
        } else {
            fromView = fromVC.view;
            toView = toVC.view;
        }
        
        UIView *containerView = [transitionContext containerView];
        [containerView addSubview:toView];
        [containerView sendSubviewToBack:toView];
        
        // 动画block
        void(^animateBlock)(void) = ^{
            fromView.frame = CGRectMake(-self.leftViewWidth, 0, self.leftViewWidth, fromView.frame.size.height);
            if (self.shouldMove) {
                toView.frame = CGRectMake(0, 0, toView.frame.size.width, toView.frame.size.height);
            }
            self.tapView.alpha = 0.f;
            
        };
        
        // 动画完成block
        void(^completeBlock)(void) = ^{
            if ([transitionContext transitionWasCancelled]) {
                [transitionContext completeTransition:NO];
            } else {
                [transitionContext completeTransition:YES];
                self.isShowLeft = NO;
                self.tapView.hidden = YES;
            }
        };
        
        if (self.interactive) {
            // 呵呵🙃
            [UIView animateWithDuration:[self transitionDuration:transitionContext] delay:0 options:UIViewAnimationOptionCurveLinear animations:^{
                animateBlock();
            } completion:^(BOOL finished) {
                completeBlock();
            }];
        } else {
            [UIView animateWithDuration:[self transitionDuration:transitionContext] animations:^{
                animateBlock();
            } completion:^(BOOL finished) {
                completeBlock();
            }];
        }
    }
    
    
    
}

#pragma mark - setter/getter方法
- (UIView *)tapView {
    if (!_tapView) {
        _tapView = [[UIView alloc] initWithFrame:self.mainVC.view.bounds];
        _tapView.backgroundColor = [UIColor colorWithWhite:0 alpha:.4f];
        _tapView.alpha = 0.f;
        UIPanGestureRecognizer *panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
        [_tapView addGestureRecognizer:panGesture];
        
        UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissLeftMenuView)];
        [_tapView addGestureRecognizer:tapGesture];
    }
    return _tapView;
}

- (UIPanGestureRecognizer *)panGesture {
    if (!_panGesture) {
        _panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
        _panGesture.delegate = self;
    }
    return _panGesture;
}

@end

File: TestLeftSliderDemoForSwift/TestLeftSliderDemoForSwift/LeftSliderManager/WHLeftSlideManager.swift
//
//  WHLeftSlideManager.swift
//  TestLeftSliderDemoForSwift
//
//  Created by 邓伟浩 on 2020/6/1.
//  Copyright © 2020 邓伟浩. All rights reserved.
//

import UIKit

/// 手势轻扫临界速度
let LSLeftSlipCriticalVelocity: CGFloat = 800
/// 左滑手势触发距离
let LSLeftSlipLeftSlipPanTriggerWidth: CGFloat = 50

open class WHLeftSlideManager: UIPercentDrivenInteractiveTransition {
    
    weak var mainVC: UIViewController!
    weak var leftVC: UIViewController!
    
    /** 点击返回的遮罩view */
    lazy var tapView: UIView = {
        let tapView = UIView(frame: self.mainVC.view.bounds)
        tapView.backgroundColor = UIColor(white: 0, alpha: 0.4);
        tapView.alpha = 0
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(pan(pan:)))
        tapView.addGestureRecognizer(panGesture)
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(dismiss))
        tapView.addGestureRecognizer(tapGesture)
        return tapView
    }()
    
    /** present or dismiss */
    var isPresent: Bool = false

    /** 是否已经显示左滑视图 */
    var isShowLeft: Bool = false

    /** 是否在交互中 */
    var interactive: Bool = false

    /** 主VC是否跟随滑动 */
    var shouldMove: Bool = false

    /** 左滑视图宽度 */
    var leftViewWidth: CGFloat = 0.0

    /** 侧滑手势 */
    lazy var panGesture: UIPanGestureRecognizer = {
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(pan(pan:)))
        panGesture.delegate = self
        return panGesture
    }()
    
    /** 左滑管理器实例 */
    public static let shared = WHLeftSlideManager()

    /**
     *    @brief    设置左滑控制器及主控制器
     *    @param    leftViewController  左侧菜单视图控制器
     *    @param    leftViewWidth  左侧菜单视图宽度
     *    @param    mainViewController  主控制器
     *    @param    shouldMove  主视图是否跟随移动
     */
    open func set(leftViewController: UIViewController, leftViewWidth: CGFloat = scaleFor(width: 310), mainViewController: UIViewController, shouldMove: Bool = true) {
        self.leftVC = leftViewController
        self.mainVC = mainViewController
        self.leftViewWidth = leftViewWidth
        self.shouldMove = shouldMove
        
        if (self.mainVC.navigationController == nil) {
            self.mainVC.view.addSubview(self.tapView)
        } else {
            self.mainVC.navigationController?.view.addSubview(self.tapView)
        }
        self.tapView.isHidden = true
        
        self.leftVC.transitioningDelegate = self
        // 侧滑手势
        self.mainVC.view.addGestureRecognizer(self.panGesture)
        
    }
    
    /** 显示菜单 */
    func show() {
        if !self.isShowLeft {
            self.leftVC.modalPresentationStyle = .fullScreen
            self.mainVC.present(self.leftVC, animated: true, completion: nil)
        } else {
            self.dismiss()
        }
    }

    /** 取消显示菜单 */
    @objc func dismiss() {
        self.leftVC.dismiss(animated: true, completion: nil)
    }
    
    //MARK:- —————— 手势处理方法 ——————
    @objc func pan(pan: UIPanGestureRecognizer) {
        // X轴偏移
        let offsetX = pan.translation(in: pan.view).x
        // X轴速度
        let velocityX = pan.velocity(in: pan.view).x
        
        var percent: CGFloat = 0
        if (self.isShowLeft) {
            // 不能超过1
            percent = CGFloat.minimum(-offsetX / self.leftViewWidth, 1)
        } else {
            percent = CGFloat.minimum(offsetX / self.leftViewWidth, 1);
        }
        
        switch pan.state {
        case .began:
            self.interactive = true
            if self.isShowLeft {
                self.dismiss()
            } else {
                self.show()
            }
            break
        case .changed:
            self.update(percent)
            break
        case .ended, .cancelled:
            self.interactive = false
            
            // 判断是否需要转场
            var shouldTransition = false
            
            // 1.present时
            // 1.1 速度正方向，>800，则正向转场
            // 1.2 速度反向时，<-800，则反向转场
            // 1.3 速度正向<800 或者 速度反向>-800， 判断percent是否大于0.5
            if !self.isShowLeft {
                if (velocityX > 0) {
                    if (velocityX > LSLeftSlipCriticalVelocity) {
                        shouldTransition = true
                    } else {
                        shouldTransition = percent > 0.5
                    }
                } else {
                    if (velocityX < -LSLeftSlipCriticalVelocity) {
                        shouldTransition = false
                    } else {
                        shouldTransition = percent > 0.5
                    }
                }
            } else {
                if (velocityX < 0) {
                    if (velocityX < -LSLeftSlipCriticalVelocity) {
                        shouldTransition = true
                    } else {
                        shouldTransition = percent > 0.5
                    }
                } else {
                    if (velocityX > LSLeftSlipCriticalVelocity) {
                        shouldTransition = false
                    } else {
                        shouldTransition = percent > 0.5
                    }
                }
            }
            
            // 2.dismiss时
            // 2.1 速度正向，<-800，则正向转场
            // 2.2 速度反向，>800，则反向转场
            // 2.3 速度正向>-800 或者 速度反向<800，判断percent是否大于0.5
            if (shouldTransition) {
                self.finish()
            } else {
                self.cancel()
            }
            break
        default:
            break
        }
    }
    
    //MARK:- —————— UIGestureRecognizerDelegate ——————
    public func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        if self.isShowLeft {
            return true
        }
        
        guard let panGesture: UIPanGestureRecognizer = gestureRecognizer as? UIPanGestureRecognizer else {
            return false
        }
        
        // 忽略起始点不在左侧触发范围内的手势
        let touchBeganX = panGesture.location(in: panGesture.view).x
        if (touchBeganX > LSLeftSlipLeftSlipPanTriggerWidth) {
            return false
        }
        
        // 忽略反向手势
        let translation = panGesture.translation(in: panGesture.view)
        if (translation.x <= 0) {
            return false
        }
        
        return true
    }
    
    

}

public func scaleFor(width: CGFloat) -> CGFloat {
    return ((width)*(UIScreen.main.bounds.size.width/375))
}

extension WHLeftSlideManager: UIViewControllerTransitioningDelegate, UIGestureRecognizerDelegate, UIViewControllerAnimatedTransitioning {
    
    //MARK:- —————— UIViewControllerTransitioningDelegate ——————
    public func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isPresent = true
        return self
    }
    
    public func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isPresent = false
        return self
    }
    
    public func interactionControllerForDismissal(using animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return self.interactive ? self : nil
    }
    
    public func interactionControllerForPresentation(using animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return self.interactive ? self : nil
    }
    
    public func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        return 0.3
    }
    
    public func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        if self.isPresent {
            
            var toView: UIView? = nil
            var fromView: UIView? = nil
            
            if let from = transitionContext.view(forKey: .from), let to = transitionContext.view(forKey: .to) {
                fromView = from
                toView = to
            } else if let toVC = transitionContext.viewController(forKey: .to), let fromVC = transitionContext.viewController(forKey: .from) {
                fromView = fromVC.view
                toView = toVC.view
            }
            
            if let fromView = fromView, let toView = toView {
                let containerView = transitionContext.containerView
                containerView.addSubview(toView)
                toView.frame = CGRect(x: -self.leftViewWidth, y: 0, width: self.leftViewWidth, height: containerView.frame.size.height)
                
                self.tapView.superview?.bringSubviewToFront(self.tapView)
                self.tapView.isHidden = false
                
                // 动画block
                let animateBlock = {
                    toView.frame = CGRect(x: 0, y: 0, width: self.leftViewWidth, height: toView.frame.size.height)
                    if (self.shouldMove) {
                        fromView.frame = CGRect(x: self.leftViewWidth, y: 0, width: fromView.frame.size.width, height: fromView.frame.size.height)
                    }
                    self.tapView.alpha = 1.0
                }
                
                // 动画完成block
                let completeBlock = {
                    if transitionContext.transitionWasCancelled {
                        transitionContext.completeTransition(false)
                    } else {
                        transitionContext.completeTransition(true)
                        containerView.addSubview(fromView)
                        containerView.sendSubviewToBack(fromView)
                        self.isShowLeft = true
                    }
                }
                
                if self.interactive {
                    UIView.animate(withDuration: self.transitionDuration(using: transitionContext), delay: 0, options: .curveLinear, animations: {
                        animateBlock()
                    }) { (finished) in
                        completeBlock()
                    }
                } else {
                    UIView.animate(withDuration: self.transitionDuration(using: transitionContext), animations: {
                        animateBlock()
                    }) { (finished) in
                        completeBlock()
                    }
                }
            }
        } else {
            var toView: UIView? = nil
            var fromView: UIView? = nil
            
            if let from = transitionContext.view(forKey: .from), let to = transitionContext.view(forKey: .to) {
                fromView = from
                toView = to
            } else if let toVC = transitionContext.viewController(forKey: .to), let fromVC = transitionContext.viewController(forKey: .from) {
                fromView = fromVC.view
                toView = toVC.view
            }
            
            if let fromView = fromView, let toView = toView {
                let containerView = transitionContext.containerView
                containerView.addSubview(toView)
                containerView.sendSubviewToBack(toView)
                
                // 动画block
                let animateBlock = {
                    fromView.frame = CGRect(x: -self.leftViewWidth, y: 0, width: self.leftViewWidth, height: fromView.frame.size.height)
                    if (self.shouldMove) {
                        toView.frame = CGRect(x: 0, y: 0, width: toView.frame.size.width, height: toView.frame.size.height)
                    }
                    self.tapView.alpha = 1.0
                }
                
                // 动画完成block
                let completeBlock = {
                    if transitionContext.transitionWasCancelled {
                        transitionContext.completeTransition(false)
                    } else {
                        transitionContext.completeTransition(true)
                        self.isShowLeft = false
                        self.tapView.isHidden = true
                    }
                }
                
                if self.interactive {
                    // 呵呵🙃
                    UIView.animate(withDuration: self.transitionDuration(using: transitionContext), delay: 0, options: .curveLinear, animations: {
                        animateBlock()
                    }) { (finished) in
                        completeBlock()
                    }
                } else {
                    UIView.animate(withDuration: self.transitionDuration(using: transitionContext), animations: {
                        animateBlock()
                    }) { (finished) in
                        completeBlock()
                    }
                }
            }
        }
    }
    
    
}
